src/server/server_family.cc

```diff
@@ -1202,66 +1207,71 @@ void ServerFamily::Save(CmdArgList args, ConnectionContext* cntx) {
-    dest->db[i] += src.db_stats[i];
-  }
+static void MergeDbSliceStats(const DbSlice::Stats& src, Metrics* dest) {
+  if (src.db_stats.size() > dest->db_stats.size())
+    dest->db_stats.resize(src.db_stats.size());
+
+  for (size_t i = 0; i < src.db_stats.size(); ++i)
+    dest->db_stats[i] += src.db_stats[i];
 
   dest->events += src.events;
   dest->small_string_bytes += src.small_string_bytes;
 }
 
 Metrics ServerFamily::GetMetrics() const {
   Metrics result;
-
   Mutex mu;
 
+  auto cmd_stat_cb = [&dest = result.cmd_stats_map](string_view name, const CmdCallStats& stat) {
+    auto& [calls, sum] = dest[string{name}];
+    calls += stat.first;
+    sum += stat.second;
+  };
+
   auto cb = [&](unsigned index, ProactorBase* pb) {
     EngineShard* shard = EngineShard::tlocal();
     ServerState* ss = ServerState::tlocal();
 
     lock_guard lk(mu);
 
-    result.uptime = time(NULL) - this->start_time_;
+    result.coordinator_stats += ss->stats;
     result.conn_stats += ss->connection_stats;
+
+    result.uptime = time(NULL) - this->start_time_;
     result.qps += uint64_t(ss->MovingSum6());
-    result.ooo_tx_transaction_cnt += ss->stats.ooo_tx_cnt;
-    result.eval_io_coordination_cnt += ss->stats.eval_io_coordination_cnt;
-    result.eval_shardlocal_coordination_cnt += ss->stats.eval_shardlocal_coordination_cnt;
-    result.tx_schedule_cancel_cnt += ss->stats.tx_schedule_cancel_cnt;
-
-    service_.mutable_registry()->MergeCallStats(
-        index, [&dest_map = result.cmd_stats_map](string_view name, const CmdCallStats& src) {
-          auto& ent = dest_map[string{name}];
-          ent.first += src.first;
-          ent.second += src.second;
-        });
 
     if (shard) {
-      MergeInto(shard->db_slice().GetStats(), &result);
-
       result.heap_used_bytes += shard->UsedMemory();
-      if (shard->tiered_storage()) {
-        result.tiered_stats += shard->tiered_storage()->GetStats();
-      }
+      MergeDbSliceStats(shard->db_slice().GetStats(), &result);
       result.shard_stats += shard->stats();
+
+      if (shard->tiered_storage())
+        result.tiered_stats += shard->tiered_storage()->GetStats();
+
       result.traverse_ttl_per_sec += shard->GetMovingSum6(EngineShard::TTL_TRAVERSE);
       result.delete_ttl_per_sec += shard->GetMovingSum6(EngineShard::TTL_DELETE);
     }
+
+    service_.mutable_registry()->MergeCallStats(index, cmd_stat_cb);
   };
 
   service_.proactor_pool().AwaitFiberOnAll(std::move(cb));
-  result.qps /= 6;  // normalize moving average stats
+
+  // Normalize moving average stats
+  result.qps /= 6;
   result.traverse_ttl_per_sec /= 6;
   result.delete_ttl_per_sec /= 6;
 
-  result.is_master = false;
-  if (ServerState::tlocal() && ServerState::tlocal()->is_master) {
-    result.is_master = true;
+  result.is_master = ServerState::tlocal() && ServerState::tlocal()->is_master;
+  if (result.is_master)
     result.replication_metrics = dfly_cmd_->GetReplicasRoleInfo();
-  }
+
+  // Update peak stats
```

Comment:
Looks good but please add a comment saying that we rely here on GetMetrics being periodically called.

Reply:
Hm actually that's not what we want, do we?

Reply:
Maybe we can make them atomic and update them from heartbeat? But then we have to partially assemble the stats 🤯